qiskit/algorithms/minimum_eigen_solvers/adapt_qaoa.py

```diff
@@ -0,0 +1,539 @@
+                "{mixer_n_qubits[check_mixer_qubits]} to the initialised"
+                "cost operator {self.num_qubits}."
+            )
+
+    @property
+    def mixer_pool(self) -> List:
+        """Creates the mixer pool if not already defined
+
+        Returns:
+            List of mixers that make up the mixer pool.
+
+        Raises:
+            AttributeError: If operator and thus num_qubits has not yet been defined.
+        """
+        if self.cost_operator is not None:
+            if self._mixer_pool is None:
+                self._mixer_pool = adapt_mixer_pool(
+                    num_qubits=self.num_qubits, pool_type=self.mixer_pool_type
+                )
+        return self._mixer_pool
+
+    @mixer_pool.setter
+    def mixer_pool(self, mixer_pool: List) -> None:
+        self._mixer_pool = mixer_pool
+
+    @property
+    def initial_state(self) -> Optional[QuantumCircuit]:
+        """Returns an optional initial state as a circuit"""
+        if self._initial_state is not None:
+            return self._initial_state
+
+        # if no initial state is passed and we know the number of qubits, then initialize it.
+        if self.num_qubits > 0:
+            initial_state = QuantumCircuit(self.num_qubits)
+            initial_state.h(range(self.num_qubits))
+            return initial_state
+        # otherwise we cannot provide a default
+        return None
+
+    @property
+    def optimal_mixer_list(self) -> List:
+        if self._optimal_mixer_list is None:
+            self._optimal_mixer_list = []
+        return self._optimal_mixer_list
+
+    @optimal_mixer_list.setter
+    def optimal_mixer_list(self, optimal_mixer_list) -> List:
+        self._optimal_mixer_list = optimal_mixer_list
+
+    @property
+    def hyperparameter_dict(self) -> Dict:
+        """Creates dictionary of hyperparameters including ansatz parameters
+
+        Returns:
+            Dictionary of hyperparameters
+        """
+        self._hyperparameter_dict = {}
+        if self._ansatz_params:
+            self._hyperparameter_dict = dict(
+                zip(self._ansatz_params, self.best_beta + self.best_gamma)
+            )
+        return self._hyperparameter_dict
+
+    @hyperparameter_dict.setter
+    def hyperparameter_dict(self, hyperparameter_dict) -> Dict:
+        self._hyperparameter_dict = hyperparameter_dict
+
+    @property
+    def cost_operator(self):
+        """Returns an operator representing the cost of the optimization problem.
+
+        Returns:
+            OperatorBase: cost operator.
+        """
+        return self._cost_operator
+
+    @property
+    def num_qubits(self) -> int:
+        if self._cost_operator is None:
+            return 0
+        return self._cost_operator.num_qubits
+
+    @property
+    def ground_state_energy(self) -> float:
+        if self._cost_operator:
+            return min(np.real(np.linalg.eig(self._cost_operator.to_matrix())[0]))
+        else:
+            return None
```

Comment:
I don't think we should have this, see the general remarks.

Reply:
@Elliot-Coupe